src/block-gen/sequencer_rollup.py

```diff
@@ -0,0 +1,85 @@
+import pickle
+from typing import List, NewType
+from block_gen import Block, Transaction, Event
+
+# A more complex rollup that allows for privilaged block
+# production. This is achieved by allowing a "sequencer"
+# to squeeze blocks into historical L1 blocks after the fact.
+
+# This expands on the "simplest possible rollup" by, in addition
+# to blocks being generated based on the first event, the sequencer
+# may submit transactions in later blocks which contain many L2 blocks.
+
+#################### Types ####################
+
+class SequencerBlock(Block):
+    # An optional field which specifies the epoch that
+    # this sequencer block is intended for.
+    target_epoch: int
+
+SequencerBatch = NewType("SequencerBatch", List[SequencerBlock])
+
+#################### Functions ####################
+
+# Generate the full rullup chain
+def generate_rollup_chain(l1_chain: List[Block], sequencer_timeout) -> List[Block]:
+    l2_chain: List[Block] = []
+    for i in range(0, len(l1_chain) - sequencer_timeout):
+        root_block = l1_chain[i]
```

Comment:
`genesis_block` normally connotes the beginning of the entire chain, but I was thinking `root_block` would be the beginning of the epoch.

Btw note that in this algorithm I assumed that you start at the root block, scan forward to get all sequencer blocks, and then construct the epoch. HOWEVER, I updated this in my latest iteration https://github.com/ethereum-optimism/optimistic-specs/pull/29 to instead scan **backwards** so the root block actually is the last possible block that could contain sequencer blocks. That change eliminated a bunch of weird edge cases that you'll notice in this code like not being able to generate the epochs for the last few L1 blocks.

Reply:
I see! Something it might be good to mention here is that while there is an epoch per L1 block, we scan every subsequent L1 block for sequencer blocks matching that epoch, and this is necessary because the sequencer might not get its transactions in (because of congestion, censorship) or because of reorgs.

Reply:
Yep! They need to be regenerated but they would not have a new epoch number. The epoch numbers are based on the L1 block numbers, so in the same way that L1 block numbers don't change when you have a reorg, rollup epoch numbers won't change when you have a reorg.

As for what happens if the sequencer is censored, that just means that the rollup chain won't have any sequencer blocks in it, but it will continue to have deposit blocks which means that we don't lose liveness

Reply:
Is there a specification/explanation/code/discussion for how the L2 block numbers derive from L1 somewhere?